Jenkinsfile

```diff
@@ -1,69 +1,81 @@
 
 def img
 
 node {
-    stage('build') {
-        checkout(scm)
-        img = buildApp(name: 'hypothesis/lms')
-    }
-
-    stage('test') {
-        hostIp = sh(script: 'facter ipaddress_eth0', returnStdout: true).trim()
-
-        postgres = docker.image('postgres:9.4').run('-P -e POSTGRES_DB=lmstest')
-        databaseUrl = "postgresql://postgres@${hostIp}:${containerPort(postgres, 5432)}/lmstest"
-
-        try {
-            testApp(image: img, runArgs: "-u root -e TEST_DATABASE_URL=${databaseUrl} -e CODECOV_TOKEN=${credentials('LMS_CODECOV_TOKEN')}") {
-                sh 'apk add build-base postgresql-dev python3-dev yarn'
-                sh 'pip3 install -q tox>=3.8.0'
+    runArgs = "-u root -e SITE_PACKAGES=true"
 
-                // Hack to work around a race condition and make sure that
-                // .tox/.tox/bin/tox gets created before any of the parallelised
-                // `make` commands below try to access it.
-                sh 'cd /var/lib/lms && tox -e py36 --notest'
+    stage("Build") {
+        checkout scm
+        img = buildApp(name: "hypothesis/lms")
+    }
 
-                sh 'cd /var/lib/lms && make -j `nproc` checkformatting lint backend-tests'
-                sh 'cd /var/lib/lms && make coverage codecov'
+    parallel failFast: true,
+    "Formatting": {
+        stage("Formatting") {
+            testApp(image: img, runArgs: runArgs) {
+                installDeps()
+                run("make checkformatting")
             }
-        } finally {
-            postgres.stop()
         }
-
-        nodeEnv = docker.image("node:10-stretch")
-        workspace = pwd()
-        nodeEnv.inside("-u root -e HOME=${workspace}") {
-            sh 'make frontend-tests'
+    },
+    "Docstrings": {
+        stage("Docstrings") {
+            testApp(image: img, runArgs: runArgs) {
+                installDeps()
+                run("make checkdocstrings")
+            }
         }
-    }
+    },
+    "Backend lint": {
+        stage("Backend lint") {
+            testApp(image: img, runArgs: runArgs) {
+                installDeps()
+                run("make backend-lint")
+            }
+        }
+    },
+    "Backend tests": {
+        stage("Backend tests") {
+            postgres = docker.image("postgres:9.4").run("-P -e POSTGRES_DB=lmstest")
+            hostIp = sh(script: 'facter ipaddress_eth0', returnStdout: true).trim()
+            containerPort = sh(script: "docker port ${postgres.id} 5432 | cut -d: -f2", returnStdout: true).trim()
+            databaseUrl = "postgresql://postgres@${hostIp}:${containerPort}/lmstest"
 
-    onlyOnMaster {
-        stage('release') {
-            releaseApp(image: img)
+            try {
+                testApp(image: img, runArgs: "${runArgs} -e TEST_DATABASE_URL=${databaseUrl} -e CODECOV_TOKEN=${credentials('LMS_CODECOV_TOKEN')}") {
+                    installDeps()
+                    run("make backend-tests coverage codecov")
```

Comment:
We run `backend-tests coverage codecov` all at once, synchronously, because `codecov` can't work until `coverage` has completed and `coverage` can't work until `backend-tests` has completed. These are just for _backend_ coverage, we don't have frontend test coverage.